src/stp/invariants/FreedomOfRobotsInvariant.cpp

```diff
@@ -0,0 +1,42 @@
+//
+// Created by jordi on 21-04-20.
+//
+
+#include "stp/invariants/FreedomOfRobotsInvariant.h"
+
+namespace rtt::ai::stp::invariant {
+
+FreedomOfRobotsInvariant::FreedomOfRobotsInvariant() noexcept {
+    /**
+     * Creates a piecewise linear function that looks as follows:
+     *
+     * (0,255)  |        XXXXXX
+     *          |       XX
+     *          |      XX
+     *          |     XX
+     *   (0,0)  |XXXXXX---------
+     *              (Distance to closest enemy robot)
+     */
+    piecewiseLinearFunction = nativeformat::param::createParam(control_constants::FUZZY_FALSE, control_constants::FUZZY_TRUE, control_constants::FUZZY_FALSE, "distanceToClosestEnemyRobot");
+    piecewiseLinearFunction->setYAtX(control_constants::FUZZY_FALSE, 0.0);
+    piecewiseLinearFunction->setYAtX(control_constants::FUZZY_FALSE, stp::control_constants::DISTANCE_TO_ROBOT_CLOSE + stp::control_constants::FUZZY_MARGIN);
+    piecewiseLinearFunction->linearRampToYAtX(control_constants::FUZZY_TRUE, stp::control_constants::DISTANCE_TO_ROBOT_FAR - stp::control_constants::FUZZY_MARGIN);
+    piecewiseLinearFunction->setYAtX(control_constants::FUZZY_TRUE, stp::control_constants::DISTANCE_TO_ROBOT_FAR - stp::control_constants::FUZZY_MARGIN);
+}
+
+uint8_t FreedomOfRobotsInvariant::metricCheck(world_new::view::WorldDataView world, const world::Field* field) const noexcept {
+    auto& us = world.getUs();
+    std::vector<uint8_t> distanceMetrics{};
+
+    std::transform(us.begin(), us.end(), std::back_inserter(distanceMetrics), [&](auto& robot) {
+        auto robotPosition = robot.get()->getPos();
+        auto distance = (world.getRobotClosestToPoint(robotPosition, world_new::them)->getPos() - robotPosition).length();
+        return calculateMetric(distance);
+    });
+
+    return std::accumulate(distanceMetrics.begin(), distanceMetrics.end(), 0) / distanceMetrics.size();
```

Comment:
Maybe calculate the average first, and use that value as parameter for calculateMetric(). This way there is only one call to the linear function instead of a maximum of 11 times.

Reply:
Ah smart

Reply:
I wonder if that works though

Reply:
I also think that wouldn't work that easily. If one robot is really really far, the average distance might be far enough but doesn't say anything about the other robots being free or not. It might work if you clamp the distances first?

Reply:
That is true actually, woops. I think if you clamp you kind of loose the advantage of calling the linear function only once. not sure tho, and seems a bit hacky to clamp that